src/webui/webapplication.cpp

```diff
@@ -914,76 +804,97 @@ void WebApplication::action_command_addCategory()
     BitTorrent::Session::instance()->addCategory(category);
 }
 
-void WebApplication::action_command_removeCategories()
+void WebApplication::removeCategories()
 {
-    CHECK_URI(0);
     CHECK_PARAMETERS("categories");
 
     QStringList categories = request().posts["categories"].split('\n');
     foreach (const QString &category, categories)
         BitTorrent::Session::instance()->removeCategory(category);
 }
 
-void WebApplication::action_command_getSavePath()
+void WebApplication::getSavePath()
 {
-    CHECK_URI(0);
     print(BitTorrent::Session::instance()->defaultSavePath());
 }
 
-bool WebApplication::isPublicScope()
+bool WebApplication::isPublicAPI(const QString &methodName) const
 {
-    return (scope_ == DEFAULT_SCOPE || scope_ == VERSION_INFO);
+    return m_publicAPIs.contains(methodName);
 }
 
 void WebApplication::doProcessRequest()
 {
-    scope_ = DEFAULT_SCOPE;
-    action_ = DEFAULT_ACTION;
-
-    parsePath();
-
-    if (args_.contains(".") || args_.contains("..")) {
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
         qDebug() << Q_FUNC_INFO << "Invalid path:" << request().path;
         status(404, "Not Found");
         return;
     }
 
-    if (!isPublicScope() && !sessionActive()) {
-        status(403, "Forbidden");
-        return;
-    }
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isPublicAPI(methodName) && !sessionActive()) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (actions_.value(scope_).value(action_) != 0) {
-        (this->*(actions_[scope_][action_]))();
+        if (!QMetaObject::invokeMethod(this, methodName.toLatin1().constData())) {
+            status(404, "Not Found");
+            qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        }
     }
     else {
-        status(404, "Not Found");
-        qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        getWebUIFile();
     }
 }
 
-void WebApplication::parsePath()
+void WebApplication::configure()
+{
+    const auto pref = Preferences::instance();
+    m_username = pref->getWebUiUsername();
+    m_password = pref->getWebUiPassword();
+
+    const QString rootFolder = Utils::Fs::expandPathAbs(
+                !pref->isAltWebUiEnabled() ? WWW_FOLDER : pref->getWebUiRootFolder());
+    if (rootFolder != m_rootFolder)
+        m_translatedFiles.clear();
+    m_rootFolder = rootFolder;
```

Comment:
should this placed under previous `if` scope?

Reply:
Fixed.